File: scripts/bindings/ggplot_discrete_null.py
"""FIX-GG04: actual nullable hue domains, mapped missing values and immutable replacement."""
import json
import sys
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes())
options=c.export_options(640,360).dpi(96).basis('current')
cases=[v for v in json.loads((ROOT/'fixtures/parity/ggplot2/discrete-null-domains.json').read_text())['cases'] if v['kind']=='hue'];records=[]
cases += json.loads((ROOT/'fixtures/parity/ggplot2/identity-null-paints.json').read_text())['cases']
def key(v):return 'Null' if v is None else {'Text':v}
def descriptor(case):
    if case.get('kind') == 'identity':
        return {'training':'Eligible','function':{'GgplotDiscreteIdentity':{'limits':None if case['limits'] is None else [key(v) for v in case['limits']],'levels':None if case['levels'] is None else [key(v) for v in case['levels']],'drop':case['drop'],'na_translate':case['na_translate'],'guide':case['guide'],'observed':[]}}}
    return {'training':'Eligible','function':{'Ordinal':{'domain':[],'range':[],'unknown':{'Explicit':None}}},'ggplot':{'Discrete':{'limits':None if case['limits'] is None else [key(v) for v in case['limits']],'levels':None if case['levels'] is None else [key(v) for v in case['levels']],'drop':case['drop'],'na_translate':case['na_translate'],'palette':{'Hue':{'h':[15,375],'chroma':100,'luminance':65,'start':0,'reverse':False}}}}}
def data_for(case):
    values=case['inputs'];return c.Data.columns({'x':c.column([float(i) for i in range(len(values))],kind='float64'),'v':c.column(values,kind='string').nullable(True)},keys=list(range(100,100+len(values))))
def build(data,case):return c.plot(data).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y(1.).color('v').color_scale('v')).scale(c.color_mapped('v',descriptor(case))).layer(c.points()).build()
def paint(v):
    if v is None:return {'red':0,'green':0,'blue':0,'alpha':0}
    if v in ('red4','blue','green','orange','NA'):
        return dict(zip(('red','green','blue','alpha'),{'red4':[139,0,0,255],'blue':[0,0,255,255],'green':[0,255,0,255],'orange':[255,165,0,255],'NA':[255,255,255,0]}[v]))
    if v=='grey50':return {'red':127,'green':127,'blue':127,'alpha':255}
    assert v.startswith('#') and len(v)==7,v
    return dict(zip(('red','green','blue','alpha'),[int(v[i:i+2],16) for i in (1,3,5)]+[255]))
def check(chart,case):
    layer=chart.semantics()['layers'][0];colors=[v['color'] for v in layer.get('styles',[])];wanted=[paint(v) for v in case['result']['training_paints'] if v is not None]
    assert colors==wanted,(case,colors,wanted)
    legend=layer.get('color_legend');entries=legend['entries'] if legend else [];domain=legend['mapping']['function'].get('Ordinal',{}).get('domain') if legend else []
    identity=case.get('kind')=='identity'
    if identity: assert len(colors)==case['result']['point_count']
    wanted_entries=[['NA' if label is None else label,paint('NA' if identity and color is None else color)] for label,color in zip(case['result']['labels'],case['result']['guide_paints'])]
    if identity and not case['guide']: wanted_entries=[]
    assert entries==wanted_entries,(case,entries,wanted_entries)
    if not identity: assert domain==[key(v) for v in case['result']['breaks']],(case,domain)
    return {'colors':colors,'entries':entries,'domain':domain}
for index,case in enumerate(cases):
    owned=[]
    try:
        data=data_for(case);owned.append(data);plot=build(data,case);owned.append(plot);wire=plot.to_json();restored=c.Plot.from_json(wire);owned.append(restored);assert restored.to_json()==wire
        chart=restored.chart();owned.append(chart);records.append({'index':index,**check(chart,case)})
        sample=None
        if case['kind']=='identity' and case['population']=='mixed' and case['levels'] is None and case['drop'] and case['na_translate'] and case['limits_name']=='first' and case['guide']:sample='identity-missing-first'
        if case['kind']=='hue' and case['population']=='mixed' and case['drop']:
            if case['levels'] is None and case['na_translate'] and case['limits_name']=='first':sample='authored-missing-first'
            if case['levels'] is not None and case['na_translate'] and case['limits_name']=='auto':sample='factor-missing-middle'
            if case['levels'] is None and not case['na_translate'] and case['limits_name']=='first':sample='suppressed-missing'
            if case['levels'] is None and not case['na_translate'] and case['limits_name']=='only':sample='empty-palette'
        if sample:
            request=output.request(restored,options);owned.append(request);frame=request.prepare();owned.append(frame)
            for fmt in ('svg','pdf','png'):(out/f'{sample}.{fmt}').write_bytes(frame.export(fmt))
            (out/f'{sample}.scene.json').write_text(json.dumps(frame.scene(),indent=2))
    finally:
        for value in reversed(owned):value.dispose()
for kind in ('hue','identity'):
    for limits_name in ('auto','first'):
        for translate in (False,True):
            selected={v['population']:v for v in cases if v['kind']==kind and (kind=='hue' or v['guide']) and v['levels'] is None and v['limits_name']==limits_name and v['na_translate']==translate and v['drop']}
            owned=[]
            try:
                original=data_for(selected['mixed']);owned.append(original);plot=build(original,selected['mixed']);owned.append(plot);wire=plot.to_json();chart=plot.chart();owned.append(chart)
                request=output.request(plot,options);owned.append(request);held=request.prepare();owned.append(held);scene=held.scene()
                for population in ('finite','missing','empty','mixed'):
                    case=selected[population];replacement=data_for(case);owned.append(replacement);updates=chart.transaction();owned.append(updates);builder=updates.replace(original,replacement);owned.append(builder);update=builder.build();owned.append(update);receipt=chart.commit(update);assert 'Applied' in receipt,receipt
                    fresh=build(replacement,case);owned.append(fresh);batch=fresh.chart();owned.append(batch);actual=check(chart,case);assert actual==check(batch,case)
                    assert held.scene()==scene and plot.to_json()==wire
                    records.append({'kind':kind,'replacement':population,'limits':limits_name,'translate':translate,**actual})
            finally:
                for value in reversed(owned):value.dispose()
position_cases=json.loads((ROOT/'fixtures/parity/ggplot2/missing-paint-positions.json').read_text())['cases']
for index,case in enumerate(position_cases):
    owned=[]
    try:
        columns={axis:c.categorical(case[axis]) if case[axis+'_kind']=='category' else c.column(case[axis],kind='float64') for axis in ('x','y')}
        columns['v']=c.column(case['values'],kind='string').nullable(True)
        data=c.Data.columns(columns,keys=list(range(100,104)));owned.append(data)
        spec=descriptor({**case,'drop':True,'levels':None});spec['guide']='Hidden'
        plot=c.plot(data).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y('y').color('v').color_scale('v')).scale(c.color_mapped('v',spec)).layer(c.points()).x_axis(c.x_axis().range(100,540)).y_axis(c.y_axis().range(300,60)).build();owned.append(plot)
        request=output.request(plot,options);owned.append(request);frame=request.prepare();owned.append(frame);scene=frame.scene();points=[None]*4
        for item,targets in zip(scene['items'],scene['targets']):
            if not targets or 'Source' not in targets[0]:continue
            point=item['primitive']['Point'];row=int(targets[0]['Source']['key'])-100;assert points[row] is None
            points[row]=[(point['center']['x']-100)/440,(300-point['center']['y'])/240]
            assert point['fill']==paint(case['result']['colors'][row])
        for row,point in enumerate(points):
            if case['result']['colors'][row] is None:assert point is None,(case,row,point)
            else:
                assert point is not None,(case,row)
                assert all(abs(point[i]-case['result'][axis+'_positions'][row])<1e-12 for i,axis in enumerate(('x','y'))),(case,row,point)
        guides=frame.guides()['guides'];labels={}
        for axis,side,start,span in [('x','Bottom',100,440),('y','Left',300,-240)]:
            guide=next(g for g in guides if g['spec']['side']==side);labels[axis]=sorted(t['label'] for t in guide['ticks']);assert labels[axis]==sorted(case['result'][axis+'_labels'])
            for label,position in zip(case['result'][axis+'_labels'],case['result'][axis+'_major_positions']):
                tick=next(t for t in guide['ticks'] if t['label']==label);assert abs((tick['position']-start)/span-position)<1e-12,(case,tick,position)
        records.append({'position_index':index,'points':points,'labels':labels})
        if case['x_kind']=='category' and case['y_kind']=='numeric' and case['population']=='mixed' and case['limits_name']=='first' and not case['na_translate']:
            for fmt in ('svg','pdf','png'):(out/f'unpainted-categories.{fmt}').write_bytes(frame.export(fmt))
    finally:
        for value in reversed(owned):value.dispose()
empty_case=next(case for case in cases if case['population']=='empty' and case['limits_name']=='auto' and case['levels'] is None)
for retained in (False,True):
    owned=[]
    try:
        data=data_for(empty_case);owned.append(data);spec=descriptor(empty_case);spec['ggplot']['Discrete']['empty_population']=retained
        plot=c.plot(data).profile('Ggplot2_4_0_3').aes(c.aes().x('x').y(1.).color('v').color_scale('v')).scale(c.color_mapped('v',spec)).layer(c.points()).build();owned.append(plot);wire=plot.to_json();encoded=json.loads(wire);version=21 if retained else 17;assert encoded['version']==version
        restored=c.Plot.from_json(wire);owned.append(restored);assert restored.to_json()==wire
        if retained:
            encoded['version']=20;rejected=False
            try:
                downgraded=c.Plot.from_json(json.dumps(encoded));owned.append(downgraded)
            except c.ChartError as error:
                assert json.loads(str(error))['code']=='CHART_UNSUPPORTED_CAPABILITY';rejected=True
            assert rejected
        else:assert 'empty_population' not in wire
        records.append({'retained_empty':retained,'version':version})
    finally:
        for value in reversed(owned):value.dispose()
assert len(records)==874,len(records)
(out/'records.json').write_text(json.dumps(records,indent=2));options.dispose();output.dispose()
print('PASS Python: 256 hue and 512 identity nullable domains, 32 replacements, 72 position panels, two wire cases and six SVG/PDF/PNG samples.')

File: crates/chart-core/tests/ggplot_discrete_null.rs
//! FIX-GG04: nullable domain order, missing-palette behavior and shared identity training.
use chart_core::{color::parse_r, interpolate::Value, prelude::*, scales::*};
use serde_json::Value as Json;
fn keys(v: &Json) -> Vec<ScaleKey> {
    v.as_array()
        .unwrap()
        .iter()
        .map(|v| {
            v.as_str()
                .map_or(ScaleKey::Null, |s| ScaleKey::Text(s.into()))
        })
        .collect()
}
fn descriptor(case: &Json) -> MappedScaleSpec {
    let limits = case["limits"].as_array().map(|_| keys(&case["limits"]));
    let levels = case["levels"].as_array().map(|_| keys(&case["levels"]));
    if case["kind"] == "identity" {
        let mut spec = MappedScaleSpec::authored(ScaleFunctionSpec::GgplotDiscreteIdentity(
            GgplotDiscreteIdentity {
                limits,
                levels,
                drop: case["drop"].as_bool().unwrap(),
                na_translate: case["na_translate"].as_bool().unwrap(),
                guide: case["guide"].as_bool().unwrap_or(true),
                observed: vec![],
            },
        ));
        spec.training = ScaleTraining::Eligible;
        spec
    } else {
        MappedScaleSpec::authored(ScaleFunctionSpec::Ordinal(OrdinalSpec::default()))
            .with_ggplot(GgplotScalePolicy::Discrete {
                empty_population: false,
                limits,
                levels,
                drop: case["drop"].as_bool().unwrap(),
                na_translate: case["na_translate"].as_bool().unwrap(),
                palette: GgplotDiscretePalette::Hue(Default::default()),
            })
            .unwrap()
    }
}
fn fixture() -> Json {
    serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/discrete-null-domains.json"
    ))
    .unwrap()
}
fn paint(value: &Json) -> chart_core::scene::Color {
    // R NA denotes absent paint; RGB channels under zero alpha are not specified.
    value.as_str().map_or(
        chart_core::scene::Color {
            red: 0,
            green: 0,
            blue: 0,
            alpha: 0,
        },
        |value| parse_r(value).unwrap().resolve(),
    )
}
#[test]
fn nullable_domains_match_reference_order_and_mapping() {
    let fixture = fixture();
    assert_eq!(fixture["cases"].as_array().unwrap().len(), 512);
    for case in fixture["cases"].as_array().unwrap() {
        assert!(case["result"].get("error").is_none(), "{case}");
        let trained = descriptor(case)
            .trained_keys(&keys(&case["inputs"]))
            .unwrap();
        let domain = match &trained.function {
            ScaleFunctionSpec::Ordinal(s) => s.domain.clone(),
            ScaleFunctionSpec::GgplotDiscreteIdentity(s) => s.domain().unwrap(),
            _ => unreachable!(),
        };
        assert_eq!(domain, keys(&case["result"]["breaks"]), "{case}");
        let wire = serde_json::to_string(&trained).unwrap();
        assert_eq!(
            serde_json::from_str::<MappedScaleSpec>(&wire).unwrap(),
            trained
        );
        if case["kind"] == "hue" {
            let mapped = MappedScale::for_colors(trained).unwrap();
            for (key, expected) in keys(&case["query"])
                .iter()
                .zip(case["result"]["mapped"].as_array().unwrap())
            {
                assert_eq!(
                    mapped
                        .color(None, Some(key), parse_r("grey50").unwrap().resolve())
                        .unwrap(),
                    paint(expected),
                    "{case}"
                );
            }
        } else {
            let mapped = MappedScale::new(trained).unwrap();
            for (key, expected) in keys(&case["query"])
                .iter()
                .zip(case["result"]["mapped"].as_array().unwrap())
            {
                assert_eq!(
                    mapped.category(Some(key)).unwrap(),
                    expected
                        .as_str()
                        .map_or(Value::Missing, |s| Value::Text(s.into())),
                    "{case}"
                );
            }
        }
    }
}
#[test]
fn primary_nullable_color_domains_preserve_labels_and_marks_through_json() {
    primary_nullable_paints(&fixture(), "hue");
}
#[test]
fn primary_identity_null_paints_preserve_raw_colors_through_json() {
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/identity-null-paints.json"
    ))
    .unwrap();
    assert_eq!(fixture["cases"].as_array().unwrap().len(), 512);
    primary_nullable_paints(&fixture, "identity");
}
fn primary_nullable_paints(fixture: &Json, kind: &str) {
    for case in fixture["cases"]
        .as_array()
        .unwrap()
        .iter()
        .filter(|c| c["kind"] == kind)
    {
        let values = case["inputs"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().map(String::from))
            .collect::<Vec<_>>();
        let data = Data::columns()
            .column("x", (0..values.len()).map(|i| i as f64).collect::<Vec<_>>())
            .column("v", values)
            .build()
            .unwrap();
        let plot = plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .aes(aes().x("x").y(1.).color("v").color_scale("v"))
            .scale(color_mapped("v", descriptor(case)))
            .layer(points())
            .build()
            .unwrap();
        let wire = plot.to_json().unwrap();
        let restored = Plot::from_json(&wire).unwrap();
        assert_eq!(restored.to_json().unwrap(), wire);
        let prepared = restored.chart().unwrap().prepare().unwrap();
        let layer = &prepared.layers()[0];
        let actual = layer
            .marks()
            .iter()
            .map(|mark| mark.style.color)
            .collect::<Vec<_>>();
        let expected = case["result"]["training_paints"]
            .as_array()
            .unwrap()
            .iter()
            .filter(|value| !value.is_null())
            .map(paint)
            .collect::<Vec<_>>();
        assert_eq!(actual, expected, "{case}");
        if let Some(count) = case["result"]["point_count"].as_u64() {
            assert_eq!(actual.len(), count as usize, "{case}");
        }
        let actual = layer
            .color_legend()
            .map(|legend| {
                legend
                    .entries
                    .iter()
                    .map(|e| e.0.clone())
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default();
        let expected = case["result"]["labels"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap_or("NA").to_owned())
            .collect::<Vec<_>>();
        let expected = if kind == "identity" && case["guide"] == false {
            vec![]
        } else {
            expected
        };
        assert_eq!(actual, expected, "{case}");
    }
}

#[test]
fn missing_paint_preserves_both_numeric_and_category_position_domains() {
    use chart_core::{
        ChartResult, Rect, ResourceId, Revision, ScaleId,
        composition::ScaleValue,
        layout::{AxisSide, LayoutRequest, layout},
        services::*,
    };
    struct Metrics;
    impl TextMeasurer for Metrics {
        fn measure(&self, r: TextRequest<'_>) -> ChartResult<TextMetrics> {
            TextMetrics::new(r.text.len() as f64 * 5., 8., 2.)
        }
    }
    let request = LayoutRequest::new(
        Rect::new(0., 0., 640., 360.).unwrap(),
        Units::LogicalPixels,
        ResourceDescriptor {
            id: ResourceId::new(1),
            revision: Revision::INITIAL,
            kind: ResourceKind::Font,
            byte_len: 1,
        },
    );
    let fixture: Json = serde_json::from_str(include_str!(
        "../../../fixtures/parity/ggplot2/missing-paint-positions.json"
    ))
    .unwrap();
    assert_eq!(fixture["cases"].as_array().unwrap().len(), 72);
    for case in fixture["cases"].as_array().unwrap() {
        let mut data = Data::columns();
        for (axis, kind) in [("x", "x_kind"), ("y", "y_kind")] {
            let values = case[axis].as_array().unwrap();
            data = if case[kind] == "category" {
                data.column(
                    axis,
                    categorical(
                        values
                            .iter()
                            .map(|v| v.as_str().unwrap().to_owned())
                            .collect::<Vec<_>>(),
                    ),
                )
            } else {
                data.column(
                    axis,
                    values
                        .iter()
                        .map(|v| v.as_f64().unwrap())
                        .collect::<Vec<_>>(),
                )
            };
        }
        let data = data
            .column(
                "v",
                case["values"]
                    .as_array()
                    .unwrap()
                    .iter()
                    .map(|v| v.as_str().map(String::from))
                    .collect::<Vec<_>>(),
            )
            .build()
            .unwrap();
        let mut policy_case = case.clone();
        policy_case["kind"] = "hue".into();
        policy_case["drop"] = true.into();
        policy_case["levels"] = Json::Null;
        let mut scale = descriptor(&policy_case);
        scale.guide = Some(Box::new(GgplotScaleGuide::Hidden));
        let p = plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .aes(aes().x("x").y("y").color("v").color_scale("v"))
            .scale(color_mapped("v", scale))
            .layer(points())
            .x_axis(x_axis().range(100., 300.))
            .y_axis(y_axis().range(100., 300.))
            .build()
            .unwrap();
        let p = Plot::from_json(&p.to_json().unwrap()).unwrap();
        let prepared = p.chart().unwrap().prepare().unwrap();
        assert_eq!(
            prepared.layers()[0].marks().len(),
            case["result"]["colors"]
                .as_array()
                .unwrap()
                .iter()
                .filter(|v| !v.is_null())
                .count(),
            "{case}"
        );
        let frame = layout(prepared, &request, &Metrics).unwrap();
        for (axis, kind, id, positions, labels, side) in [
            (
                "x",
                "x_kind",
                0,
                "x_positions",
                "x_labels",
                AxisSide::Bottom,
            ),
            ("y", "y_kind", 1, "y_positions", "y_labels", AxisSide::Left),
        ] {
            let resolved = &frame.axes()[&ScaleId::new(id)];
            for (input, expected) in case[axis]
                .as_array()
                .unwrap()
                .iter()
                .zip(case["result"][positions].as_array().unwrap())
            {
                let value = if case[kind] == "category" {
                    ScaleValue::Category(input.as_str().unwrap().into())
                } else {
                    ScaleValue::Number(input.as_f64().unwrap())
                };
                let actual = (resolved.map_value(&value).unwrap().unwrap() - 100.) / 200.;
                assert!(
                    (actual - expected.as_f64().unwrap()).abs() < 1e-12,
                    "{actual} != {expected}: {case}"
                );
            }
            let snapshots = frame.guide_snapshots();
            let guide = snapshots.iter().find(|g| g.spec.side == side).unwrap();
            assert_eq!(
                guide
                    .ticks
                    .iter()
                    .map(|t| t.label.as_str())
                    .collect::<Vec<_>>(),
                case["result"][labels]
                    .as_array()
                    .unwrap()
                    .iter()
                    .map(|v| v.as_str().unwrap())
                    .collect::<Vec<_>>(),
                "{case}"
            );
        }
    }
}

#[test]
fn retained_empty_population_requires_version_21_and_ordinary_policies_stay_17() {
    let fixture = fixture();
    let case = fixture["cases"]
        .as_array()
        .unwrap()
        .iter()
        .find(|case| {
            case["kind"] == "hue"
                && case["population"] == "empty"
                && case["limits_name"] == "auto"
                && case["levels"].is_null()
        })
        .unwrap();
    for retained in [false, true] {
        let scale = if retained {
            descriptor(case).trained_keys(&[]).unwrap()
        } else {
            descriptor(case)
        };
        let data = Data::columns()
            .column("x", Vec::<f64>::new())
            .column("v", Vec::<Option<String>>::new())
            .build()
            .unwrap();
        let p = plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .aes(aes().x("x").y(1.).color("v").color_scale("v"))
            .scale(color_mapped("v", scale))
            .layer(points())
            .build()
            .unwrap();
        let wire = p.to_json().unwrap();
        assert_eq!(
            p.definition().wire_version(),
            if retained { 21 } else { 17 }
        );
        assert_eq!(Plot::from_json(&wire).unwrap().to_json().unwrap(), wire);
        if retained {
            let mut downgraded: Json = serde_json::from_str(&wire).unwrap();
            downgraded["version"] = 20.into();
            assert!(Plot::from_json(&downgraded.to_string()).is_err());
            assert!(
                chart_core::portable::ChartEnvelope {
                    version: 20,
                    definition: p.definition().clone()
                }
                .validate()
                .is_err()
            );
        } else {
            assert!(!wire.contains("empty_population"));
        }
    }
}
